IAM 978 User invite - send email for identity creation with recovery code and link

BarcoMasile opened the pull request:
## Description
This PR introduces a new dependency for the low level email operations.
With this PR when a new identity is created through the REST API, an email is sent to the identity email with a link and code to use for the first login to complete the registration resetting the password.

Right now we're using a temporary template I made, but @huwshimi will implement a new one as soon as our UX expert provides one.

### Testing
To perform a manual test for this PR you can deploy the Admin UI with its dependencies (you can disable authentication and authorization to make testing smoother) and with the MAIL_HOST, MAIL_PORT, MAIL_USERNAME, MAIL_PASSWORD env variables, and just create an identity.
Suggestion: use `maildev` for testing, run it as a docker container with
```
$ docker run -p 1080:1080 -p 1026:1025 maildev/maildev
```

then go to `http://localhost:1080` to visit maildev dashboard. You'll be able to see any email "sent" by admin ui in that dashboard.

shipperizer commented on pkg/identities/service.go:
```diff
@@ -163,9 +168,55 @@ func (s *Service) CreateIdentity(ctx context.Context, bodyID *kClient.CreateIden
+		return err
+	}
+
+	emailAddress := identity.Traits.(map[string]interface{})["email"].(string)
```
```suggestion
	if e, ok := identity.Traits.(map[string]interface{})["email"]; ok {
emailAddress = e.(string)
}
```
reason is that the schema might have email under something named different I guess

BarcoMasile replied:
Mmmm this could be a problem since we need a way to obtain that address. I'll apply your suggestion and let's talk at today's daily about this. Thanks

shipperizer commented on pkg/identities/service.go:
```diff
@@ -163,9 +168,55 @@ func (s *Service) CreateIdentity(ctx context.Context, bodyID *kClient.CreateIden
+		RecoveryCode: code,
+	}
+
+	err = s.email.Send(ctx, emailAddress, userCreationEmailSubject, template, userCreationInviteArgs)
```
here we should chuck a context.WithTimeout and deal with the failure gracefully

BarcoMasile replied:
Right now I'm leveraging whatever timeout/deadline the request context has.
Should we add this in this PR or can we consider it an enhancement for another time?

shipperizer replied:
server timeout is probably 60s so it would be a problem 

add an issue in the repo and we ll tackle it next iteration 

fix should be quick nonetheless

nsklikas commented on pkg/identities/service.go:
```diff
@@ -163,9 +168,62 @@ func (s *Service) CreateIdentity(ctx context.Context, bodyID *kClient.CreateIden
+		RecoveryCode: code,
+	}
+
+	err = s.email.Send(ctx, emailAddress, userCreationEmailSubject, template, userCreationInviteArgs)
```
If this fails the identity will be created, but we will return an error. This means that the user will be unable to input the same username when the error is fixed, but they will also be unable to access this account.

BarcoMasile replied:
Yes, in this case the admin will have to remove the identity and create it again.
We can improve on this in the future

nsklikas reviewed (commented):
Good job overall, I have a couple of comments. 

This breaks the skaffold deployment, will this be addressed in another PR?

On top of that I had trouble testing it locally. I was unable to make the maildev server to work. I started the skaffold setup, then spun up a localhost admin-ui and configured it with:
```
"MAIL_HOST": "0.0.0.0",
"MAIL_PORT": "1025",
"MAIL_FROM_ADDRESS": "a@b.com",
```
I couldn't use the API via the frontend, because the validation failed as the frontend does not allow you to set a password. I had to use curl to access the API:
```
curl 'http://localhost:8080/api/v0/identities' -X POST \
  -H 'Accept: application/json, text/plain, */*' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -H 'Cookie: <cookie>' \
  -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' \
  --data-raw '{
  "credentials": {"password": {"config": {"password": "123"}}},
  "schema_id": "default",
  "traits": {
    "email": "a@s.com"
  }
}'
```
But this resulted in the following panic:
```
2024/09/09 15:19:39 http: panic serving 127.0.0.1:41704: runtime error: invalid memory address or nil pointer dereference
goroutine 245 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1903 +0x13f
panic({0x27c6360?, 0x3ca4d80?})
	/usr/local/go/src/runtime/panic.go:770 +0x136
github.com/canonical/identity-platform-admin-ui/pkg/identities.(*API).error(0xc0007e4060, 0x0)
	/home/nikos/projects/identity-platform-admin-ui/pkg/identities/handlers.go:259 +0x43
github.com/canonical/identity-platform-admin-ui/pkg/identities.(*API).handleCreate(0xc0007e4060, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00088e5a0)
	/home/nikos/projects/identity-platform-admin-ui/pkg/identities/handlers.go:155 +0x608
net/http.HandlerFunc.ServeHTTP(0xc00022da50, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00088e5a0)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/canonical/identity-platform-admin-ui/internal/validation.(*ValidationRegistry).ValidationMiddleware.func1({0x75a1e1f812f8, 0xc0008eb700}, 0xc0007f4a20)
	/home/nikos/projects/identity-platform-admin-ui/internal/validation/registry.go:89 +0xa74
net/http.HandlerFunc.ServeHTTP(0xc00052e380, {0x75a1e1f812f8, 0xc0008eb700}, 0xc0007f4900)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/canonical/identity-platform-admin-ui/internal/authorization.(*Middleware).Authorize.func1.1({0x75a1e1f812f8, 0xc0008eb700}, 0xc0007f4120)
	/home/nikos/projects/identity-platform-admin-ui/internal/authorization/middlewares.go:161 +0x67b
net/http.HandlerFunc.ServeHTTP(0xc00052e3a0, {0x75a1e1f812f8, 0xc0008eb700}, 0xc0007f4120)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/canonical/identity-platform-admin-ui/pkg/authentication.(*Middleware).oAuth2CookieAuthentication.func1({0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f7a0)
	/home/nikos/projects/identity-platform-admin-ui/pkg/authentication/middlewares.go:139 +0x9d3
net/http.HandlerFunc.ServeHTTP(0xc00052e3c0, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f7a0)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/canonical/identity-platform-admin-ui/pkg/authentication.(*Middleware).oAuth2BearerAuthentication.func1({0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f680)
	/home/nikos/projects/identity-platform-admin-ui/pkg/authentication/middlewares.go:86 +0x59b
net/http.HandlerFunc.ServeHTTP(0xc00052e3e0, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f560)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/go-chi/chi/v5.(*ChainHandler).ServeHTTP(0xc000896600, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f560)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/chi/v5/chain.go:31 +0x34
github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0006f9f20, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f560)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/chi/v5/mux.go:459 +0x306
net/http.HandlerFunc.ServeHTTP(0xc0008284a0, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00003f560)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/go-chi/chi/v5/middleware.RequestLogger.func1.1({0x75a1e1f812f8, 0xc0008eb680}, 0xc00003ea20)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/chi/v5/middleware/logger.go:55 +0x28f
net/http.HandlerFunc.ServeHTTP(0xc000804c30, {0x75a1e1f812f8, 0xc0008eb680}, 0xc00003ea20)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/go-chi/cors.(*Cors).Handler.func1({0x75a1e1f812f8, 0xc0008eb680}, 0xc00003ea20)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/cors/cors.go:228 +0x2e5
net/http.HandlerFunc.ServeHTTP(0xc00067c560, {0x75a1e1f812f8, 0xc0008eb680}, 0xc00003ea20)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/canonical/identity-platform-admin-ui/internal/monitoring.(*Middleware).ResponseTime.func1.1({0x2e44690, 0xc0007d4c00}, 0xc00003ea20)
	/home/nikos/projects/identity-platform-admin-ui/internal/monitoring/middlewares.go:39 +0x127
net/http.HandlerFunc.ServeHTTP(0xc00067c580, {0x2e44690, 0xc0007d4c00}, 0xc00003ea20)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/go-chi/chi/v5/middleware.RequestID.func1({0x2e44690, 0xc0007d4c00}, 0xc00003e900)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/chi/v5/middleware/request_id.go:76 +0x253
net/http.HandlerFunc.ServeHTTP(0xc0005dca08, {0x2e44690, 0xc0007d4c00}, 0xc00003e900)
	/usr/local/go/src/net/http/server.go:2171 +0x33
github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0006f9f20, {0x2e44690, 0xc0007d4c00}, 0xc00003e900)
	/home/nikos/projects/identity-platform-admin-ui/vendor/github.com/go-chi/chi/v5/mux.go:90 +0x29f
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP(0xc0008ce0b0, {0x2e44690, 0xc0007d4c00}, 0xc00003cb40, {0x2e2f480, 0xc0006f9f20})
	/home/nikos/projects/identity-platform-admin-ui/vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/handler.go:217 +0x1468
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1({0x2e400b8, 0xc0008fe000}, 0xc00003cb40)
	/home/nikos/projects/identity-platform-admin-ui/vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/handler.go:81 +0x51
net/http.HandlerFunc.ServeHTTP(0xc0008d8400, {0x2e400b8, 0xc0008fe000}, 0xc00003cb40)
	/usr/local/go/src/net/http/server.go:2171 +0x33
net/http.serverHandler.ServeHTTP({0xc0008e4000}, {0x2e400b8, 0xc0008fe000}, 0xc00003cb40)
	/usr/local/go/src/net/http/server.go:3142 +0x257
net/http.(*conn).serve(0xc0008dc360, {0x2e4db10, 0xc0008c4960})
	/usr/local/go/src/net/http/server.go:2044 +0x1ab5
created by net/http.(*Server).Serve in goroutine 242
	/usr/local/go/src/net/http/server.go:3290 +0xa9a
```

The reason for this is that:
```
ids, err := a.service.CreateIdentity(r.Context(), &identity.CreateIdentityBody)
```
returned:
```"dial failed: dial tcp 0.0.0.0:1025: connect: connection refused"```

BarcoMasile commented on internal/mail/service.go:
```diff
@@ -0,0 +1,98 @@
+
+	// treat smtp connection as authenticated only if username is passed
+	if username != "" {
+		mailOpts = []mail.Option{mail.WithSMTPAuth(mail.SMTPAuthPlain), mail.WithUsername(username), mail.WithPassword(password)}
```
Yes, this embarassingly slipped. Thanks for picking it up!

BarcoMasile commented on internal/mail/service.go:
```diff
@@ -0,0 +1,98 @@
+	return e.client.DialAndSendWithContext(ctx, msg)
+}
+
+func NewEmailService(host string, port int, username, password, from string, tracer trace.Tracer, monitor monitoring.MonitorInterface, logger logging.LoggerInterface) *EmailService {
```
I like that too, but @shipperizer said in a previous PR he prefers it if we do it like this.
I'm open to changing it

nsklikas replied:
Huh, I thought we had this pattern in other places as well. No problem with keeping it as it is, if that's what the team prefers.

shipperizer replied:
I think rule should be no more than 4/5 args, after that point we go with aggregate objects

BarcoMasile replied:
Ok nice, I'll change it to the config object then
Thanks guys

BarcoMasile commented on pkg/web/router.go:
```diff
@@ -164,6 +170,7 @@ func NewRouter(config *RouterConfig, wpool pool.WorkerPoolInterface) http.Handle
 			"/api/v0/auth/callback",
 			"/api/v0/status",
 			"/api/v0/metrics",
+			"/test",
```
This slipped also, sorry and thanks for pointing it out

BarcoMasile commented:
> This breaks the skaffold deployment, will this be addressed in another PR?
> 
Yes, we did it also for the auth implementation, we'll ask our trusty @shipperizer to adjust the skaffold :D

> On top of that I had trouble testing it locally. I was unable to make the maildev server to work. I started the skaffold setup, then spun up a localhost admin-ui and configured it with:
> 
> ```
> "MAIL_HOST": "0.0.0.0",
> "MAIL_PORT": "1025",
> "MAIL_FROM_ADDRESS": "a@b.com",
> ```
For local testing, I used `localhost` as `MAIL_HOST`. Try that also.


> I couldn't use the API via the frontend, because the validation failed as the frontend does not allow you to set a password. I had to use curl to access the API:
> 

I assume this needs to be addressed by the web team to fix it on the frontend.


> ```
> 2024/09/09 15:19:39 http: panic serving 127.0.0.1:41704: runtime error: invalid memory address or nil pointer dereference
> goroutine 245 [running]:
> net/http.(*conn).serve.func1()
> 	/usr/local/go/src/net/http/server.go:1903 +0x13f
> panic({0x27c6360?, 0x3ca4d80?})
> 	/usr/local/go/src/runtime/panic.go:770 +0x136
> github.com/canonical/identity-platform-admin-ui/pkg/identities.(*API).error(0xc0007e4060, 0x0)
> 	/home/nikos/projects/identity-platform-admin-ui/pkg/identities/handlers.go:259 +0x43
> github.com/canonical/identity-platform-admin-ui/pkg/identities.(*API).handleCreate(0xc0007e4060, {0x75a1e1f812f8, 0xc0008eb700}, 0xc00088e5a0)
> 	/home/nikos/projects/identity-platform-admin-ui/pkg/identities/handlers.go:155 +0x608
 
> The reason for this is that:
> 
> ```
> ids, err := a.service.CreateIdentity(r.Context(), &identity.CreateIdentityBody)
> ```
> 
> returned: `"dial failed: dial tcp 0.0.0.0:1025: connect: connection refused"`

About this, it is actually 2 different issues:
- the CreateIdentity returns an error, and that's related to the maildev container, make sure it's exposed on the right port and use `localhost` as MAIL_HOST.

The other issue is a null pointer exception in the `error` method inside the identities handlers.
It seems to lament that the `*kClient.GenericError` passed in the method is actually nil.
On this one I can open an issue. Thing is, I didn't change how that value gets passed so that bug must have been in there already.

wood-push-melon commented:
Throw some dummy questions because I probably lack some context about the feature:

- What SMTP service will be used when going to staging/production?
- Did not find the corresponding CLI command. Is it going to be another PR?
- A very minor stuff: should we add the new env vars to the [`README.md`](https://github.com/canonical/identity-platform-admin-ui?tab=readme-ov-file#environment-variables)? (Maybe we would like to find another way to doc the env vars in the future)

wood-push-melon commented on internal/config/specs.go:
```diff
@@ -57,4 +57,10 @@ type EnvSpec struct {
 
 	OpenFGAWorkersTotal int `envconfig:"openfga_workers_total" default:"150"`
+
+	MailHost        string `envconfig:"MAIL_HOST" required:"true"`
```
Just have a question about the new env vars:

When it comes to the admin UI charm, how do we like to pass the values to these env vars? From the integration data (e.g. integration with a certain SMTP charm) or from juju configs?

BarcoMasile replied:
I was thinking we would do it trhough integration data with an smtp charm
What do you think @shipperizer ?

shipperizer replied:
yes i think integration is the right way to go

nsklikas commented:
I didn't notice that you were forwarding the mail server's port to `1026`, that's why it didn't work before. Now it works great.

nsklikas commented:
> About this, it is actually 2 different issues:
> 
>     * the CreateIdentity returns an error, and that's related to the maildev container, make sure it's exposed on the right port and use `localhost` as MAIL_HOST.
> 
> 
> The other issue is a null pointer exception in the `error` method inside the identities handlers. It seems to lament that the `*kClient.GenericError` passed in the method is actually nil. On this one I can open an issue. Thing is, I didn't change how that value gets passed so that bug must have been in there already.

Before this change, the `Service.CreateIdentity` method would return an error only if an error was returned from Kratos. In that case the Error field in the Kratos response would always be populated. With this change, an error can occur from the mail server as well. This causes this panic, as the Kratos response (correctly) has no error.

shipperizer reviewed (commented):
I think we need a service/backoffice endpoint that is only exposed internally just in case stuff goes wrong and we need to resend the email

happy for an issue to be created and tackled down the line

BarcoMasile commented:
> Before this change, the `Service.CreateIdentity` method would return an error only if an error was returned from Kratos. In that case the Error field in the Kratos response would always be populated. With this change, an error can occur from the mail server as well. This causes this panic, as the Kratos response (correctly) has no error.

@nsklikas I fixed this by separating at the service level the creation of the identity from the sending of the email (I should have done that earlier so each service method only has a single responsibility.

> I think we need a service/backoffice endpoint that is only exposed internally just in case stuff goes wrong and we need to resend the email
>
> happy for an issue to be created and tackled down the line

@shipperizer I'll add a handler for sending the email. I'll either add the timeout or open an issue for it.

shipperizer commented:
> @shipperizer I'll add a handler for sending the email.

cli command is probably better as login UI has no authorization

nsklikas reviewed (approved):
Looks much better now, good job

BarcoMasile commented:
@wood-push-melon 
> * What SMTP service will be used when going to staging/production?
This is up to who deploys the platform, I personally don't have a strong preference

> * Did not find the corresponding CLI command. Is it going to be another PR?
After talkin to @shipperizer , this should be part of [IAM-979](https://warthogs.atlassian.net/browse/IAM-979?atlOrigin=eyJpIjoiOWE2MzU1MWY0NGJjNDc0YmEzMDE4YWMxMDlhMDc3MDQiLCJwIjoiaiJ9) (I will need to update that description though)

> * A very minor stuff: should we add the new env vars to the [`README.md`](https://github.com/canonical/identity-platform-admin-ui?tab=readme-ov-file#environment-variables)? (Maybe we would like to find another way to doc the env vars in the future)
Yes, we should! I forgot about that, I was a little in a rush to push this PR, my bad. Thanks for reminding me!

BarcoMasile commented:
@shipperizer 
> cli command is probably better as login UI has no authorization

I'll add a task for that then since it deserves it's own activity

BarcoMasile commented:
@shipperizer @nsklikas @wood-push-melon 
Added env vars to the Readme, added a new env var set an explicit timeout on the sending of the mail, default value 15 seconds.
And switched to the compound obeject for the email service creation.

shipperizer commented:
@BarcoMasile create a ticket for skaffold fixes

good to merge

BarcoMasile commented:
- https://github.com/canonical/identity-platform-admin-ui/issues/408
- https://github.com/canonical/identity-platform-admin-ui/issues/409
